src/engine/BoundManager.cpp

```diff
@@ -296,3 +304,213 @@ void BoundManager::registerRowBoundTightener( IRowBoundTightener *ptrRowBoundTig
     ASSERT( _rowBoundTightener == nullptr );
     _rowBoundTightener = ptrRowBoundTightener;
 }
+
+void BoundManager::explainBound( unsigned variable, bool isUpper, Vector<double> &explanation ) const
+{
+     ASSERT( shouldProduceProofs() && variable < _size );
+    auto temp = _boundExplainer->getExplanation( variable, isUpper );
+    explanation = Vector<double>( temp.size() );
+    for ( unsigned i = 0; i < temp.size(); ++i )
+        explanation[i] = *temp[i];
+}
+
+bool BoundManager::tightenLowerBound( unsigned variable, double value, const TableauRow &row )
+{
+    bool tightened = setLowerBound( variable, value );
+
+    if ( tightened )
+    {
+        if ( shouldProduceProofs() )
+            _boundExplainer->updateBoundExplanation( row, false, variable );
+
+        if ( _tableau != nullptr )
+            _tableau->updateVariableToComplyWithLowerBoundUpdate( variable, value );
+    }
+    return tightened;
+}
+
+bool BoundManager::tightenUpperBound( unsigned variable, double value, const TableauRow &row )
+{
+    bool tightened = setUpperBound( variable, value );
+
+    if ( tightened )
+    {
+        if ( shouldProduceProofs() )
+            _boundExplainer->updateBoundExplanation( row, true, variable );
+
+        if ( _tableau != nullptr )
+            _tableau->updateVariableToComplyWithUpperBoundUpdate( variable, value );
+    }
+    return tightened;
+}
+
+bool BoundManager::tightenLowerBound( unsigned variable, double value, const SparseUnsortedList &row )
+{
+    bool tightened = setLowerBound( variable, value );
+
+    if ( tightened )
+    {
+        if ( shouldProduceProofs() )
+            _boundExplainer->updateBoundExplanationSparse( row, false, variable );
+
+        if ( _tableau != nullptr )
+            _tableau->updateVariableToComplyWithLowerBoundUpdate( variable, value );
+    }
+    return tightened;
+}
+
+bool BoundManager::tightenUpperBound( unsigned variable, double value, const SparseUnsortedList &row )
+{
+    bool tightened = setUpperBound( variable, value );
+
+    if ( tightened )
+    {
+        if ( shouldProduceProofs() )
+            _boundExplainer->updateBoundExplanationSparse( row, true, variable );
+
+        if ( _tableau != nullptr )
+            _tableau->updateVariableToComplyWithUpperBoundUpdate( variable, value );
+    }
+
+    return tightened;
+}
+
+void BoundManager::resetExplanation( const unsigned var, const bool isUpper ) const
+{
+    _boundExplainer->resetExplanation( var, isUpper );
+}
+
+void BoundManager::setExplanation( const Vector<double> &explanation, unsigned var, bool isUpper ) const
+{
+    ASSERT( explanation.size() == _boundExplainer->getNumberOfRows() || explanation.empty() );
+    _boundExplainer->setExplanation( explanation, var, isUpper );
+}
+
+BoundExplainer *BoundManager::getBoundExplainer() const
```

Comment:
The use of "const" here is a bit deceptive, because we're returning a pointer to an internal member that can later be changed. If the code allows, better to add another const:

const BoundExplainer *BoundManager::getBoundExplainer() const

To signal that the member also cannot be changed later